### PR TITLE
feat: `pool.recycleWorkers()` method

### LINKED ISSUE #62: Feature: task level isolation
Vitest will need "task level" worker isolation for following features and fixes:

- https://github.com/vitest-dev/vitest/issues/3659
- https://github.com/vitest-dev/vitest/pull/3203
- https://github.com/vitest-dev/vitest/pull/3491

Naming things is hard, but by "task level isolation" I mean a situation where pool's `isolateWorkers` can be set as `false`, but then pool itself can be instructed to use fresh new workers for tasks on demand. There are situations where we may not need isolation at all, but based on dynamic variables on runtime we'll suddenly need to recycle workers for next tasks. 

```js
import { Tinypool } from "tinypool";

const pool = new Tinypool({
  // Workers are not isolated at "pool level"
  isolateWorkers: false,
  ...
});

// Task takes "environment: 'jsdom' | 'node'" as argument
const options = { environment: "jsdom" };

await pool.run("taskName", options) // workerId #1
await pool.run("taskName", options) // workerId #1
await pool.run("taskName", options) // workerId #1

// Change task options. Next tasks need to be run in new workers even though "pool level" isolation is disabled
options.environment = "node";

// Indicate pool to use new worker for next tasks. Could be something like:
// Imperative method "flushWorkers":
await pool.flushWorkers();
// Or new argument in the pool.run() method, e.g. "runInNewWorker":
options.runInNewWorker = true

// Eventually new worker should be used for next tasks and workerId should change
await pool.run("taskName", options) // workerId #2
await pool.run("taskName", options) // workerId #2
await pool.run("taskName", options) // workerId #2
```

Or maybe a `isolationKey` which would be used to see if given tasks can share the worker:

```js
options.environment = "jsdom"; // Option for task
options.isolationKey = "some-key-1"; // Tinypool's option
await pool.run("taskName", options) // workerId #1
await pool.run("taskName", options) // workerId #1
await pool.run("taskName", options) // workerId #1

options.environment = "node"; // Option for task
// Worker ID is same as isolationKey did not change:
await pool.run("taskName", options) // workerId #1

options.isolationKey = "some-key-2"; // Tinypool's option
await pool.run("taskName", options) // workerId #2        <---
await pool.run("taskName", options) // workerId #2        <---
await pool.run("taskName", options) // workerId #2        <---
```

Some pitfalls to consider:
- If there are on-going tasks, how would `pool.flushWorkers` instruct those ones to terminate once they have resolved?
- When `options.environment` changes we only want to recycle the existing threads and not the new ones. Here the `runInNewWorker` would keep recycling all new threads which would be slow.
- `isolationKey` might be easiest to implement and handle

Let's start by looking into `pool.flushWorkers()`.

### PR DESCRIPTION
**Pull request by AriPerkkio**
- Closes #62

### CODE REVIEW

File: test/isolation.test.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+  await pool.run({})
+  expect(getThreadId()).toBe(initialThreadId)
+
+  await pool.run({})
+  await pool.run({})
+  await pool.run({})
+  expect(getThreadId()).toBe(initialThreadId)
+
+  // TODO: pool.flushWorkers()
```
Comment (Aslemammad) [inline test/isolation.test.ts:30]: @AriPerkkio  Have you checked my latest comment in the issue?

Comment (AriPerkkio) [reply test/isolation.test.ts:30]: The `filter` API is also a good idea. I answered more to the issue.

**Comment by AriPerkkio** [comment]
@sheremet-va could you test if this fixes your PR on Vitest's side?

I quickly tested this with my reproduction case from https://github.com/vitest-dev/vitest/issues/3659 and it seems to work nicely. No perf issues noticable.

```diff
        // always run environments isolated between each other
        for (const env of envs) {
+         if (envs.indexOf(env) !== 0)
+           await pool.recycleWorkers()
```

**Comment by sheremet-va** [comment]
> ```diff
> +         if (envs.indexOf(env) !== 0)
> ```

Shouldn't it always recycle? There is a possibility that some tests were running in multithreaded, so there are already some workers running (if you have a workspace project with `isolate: false` and another with `isolate: true`)

**Comment by AriPerkkio** [comment]
> Shouldn't it always recycle?

My reproduction case was simple enough that it did not create conflicts there. But sure, it's OK to set pool to recycle when ever needed. It's not that slow to create new set of workers.

**Comment by sheremet-va** [comment]
> @sheremet-va could you test if this fixes your PR on Vitest's side?

Can confirm that tests in vitest-dev/vitest#3491 are passing

**Comment by AriPerkkio** [comment]
> We also need a test for this. I think we already do, but looks like it's not correct

https://github.com/vitest-dev/vitest/commit/7d8413a7023d0ac890ed0c03f21de96c99de1a23

File: test/isolation.test.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  expect(pool.threads).toHaveLength(4)
+  const initialThreadIds = getThreadIds()
+
+  const tasks = [
+    ...times(2)(() => pool.run({ time: 500 })),
+    ...times(2)(() => pool.run({ time: 2000 })),
+  ]
+
+  await Promise.all(tasks.slice(0, 2))
+  await pool.recycleWorkers()
```
Comment (Aslemammad) [inline test/isolation.test.ts:57]: LGTM! Can we here have an expect for if the first two workers threadId changes here and the initialThreadIds don't have them anymore?

**Review by Aslemammad — CHANGES_REQUESTED** [review]

File: test/isolation.test.ts (hunk 2 above)
Comment (AriPerkkio) [reply test/isolation.test.ts:59]: Added checks here and some comments around them.

**Review by Aslemammad — APPROVED** [review]

**Pull request merged**